server_test.go

```diff
@@ -321,36 +347,88 @@ func TestOpenStatRace(t *testing.T) {
 	defer client.Close()
 	defer server.Close()
 
+	ch := make(chan result, 3)
+
 	// openpacket finishes to fast to trigger race in tests
 	// need to add a small sleep on server to openpackets somehow
 	tmppath := path.Join(os.TempDir(), "stat_race")
-	pflags := flags(os.O_RDWR | os.O_CREATE | os.O_TRUNC)
-	ch := make(chan result, 3)
+	defer os.Remove(tmppath)
+
 	id1 := client.nextID()
-	client.dispatchRequest(ch, &sshFxpOpenPacket{
-		ID:     id1,
-		Path:   tmppath,
-		Pflags: pflags,
+	client.dispatchPacket(ch, id1, &sshfx.OpenPacket{
+		Filename: tmppath,
+		PFlags:   flags(os.O_RDWR | os.O_CREATE | os.O_TRUNC),
 	})
+
 	id2 := client.nextID()
-	client.dispatchRequest(ch, &sshFxpLstatPacket{
-		ID:   id2,
+	client.dispatchPacket(ch, id2, &sshfx.LStatPacket{
 		Path: tmppath,
 	})
-	testreply := func(id uint32) {
-		r := <-ch
-		switch r.typ {
-		case sshFxpAttrs, sshFxpHandle: // ignore
-		case sshFxpStatus:
-			err := normaliseError(unmarshalStatus(id, r.data))
-			assert.NoError(t, err, "race hit, stat before open")
-		default:
-			t.Fatal("unexpected type:", r.typ)
+
+	r := <-ch
+	if r.err != nil {
+		t.Fatal("unexpected error:", r.err)
+	}
```

Comment:
Since I changed entirely how this response handling code was written, this test had to be pretty much rewritten from the ground up.

I think I’ve made the test case much more strict, while still being properly behaved. Hopefully that will make it more reliable? (There is of course still the caveat on needing to add a time delay on the `Open` so…)